### ISSUE TITLE
Disable syntax checking on parts of file when mixing Shell and Ruby

### ISSUE THREAD

**Issue by amaltson** [bbatsov/rubocop#3493]
I'm working on some scripts that mix shell and Ruby together. I'm doing this [to find a Ruby that's not a the PATH](http://stackoverflow.com/a/39283695/238292). Using a cool feature in Ruby, the `-x` flag, I have something that looks as follows:

```ruby
#!/bin/sh

export PATH=<somewhere where I think Ruby is installed>:$PATH
exec ruby -x "$0" "$@"

#!ruby

first_arg = ARGV[0]
second_arg = ARGV[1]
# more regular Ruby code here
```

When Rubocop analysis this file, it correctly finds a whole bunch of syntax errors.

```
<file>:3:18: E: unknown regexp options: ...
(Using Ruby 2.1 parser; configure using TargetRubyVersion parameter, under AllCops)
export PATH=<standard path location>:$PATH
                 ^^^^^^
```

I've tried ignoring AllCops on the shell lines like so:

```ruby
#!/bin/sh

# rubocop:disable all
export PATH=<somewhere where I think Ruby is installed>:$PATH
exec ruby -x "$0" "$@"

#!ruby
# rubocop:enable all

first_arg = ARGV[0]
second_arg = ARGV[1]
# more regular Ruby code here
```

But Rubocop still errors out because this is a syntax error, not a warning.

--------

## Expected behavior

I guess I was expecting the disable block to also ignore syntax errors.

## Actual behavior

Disable block is ignored when the code has syntax errors.

## Steps to reproduce the problem

See above description, but here's a block of code:

```ruby
#!/bin/sh

# rubocop:disable all
export PATH=/usr/local/bin:/usr/bin:$PATH
exec ruby -x "$0" "$@"

#!ruby
# rubocop:enable all

puts ARGV[0]
# more regular Ruby code here
```

## RuboCop version

Include the output of `rubocop -V`:

```
$ bundle exec rubocop -V
0.42.0 (using Parser 2.3.1.2, running on ruby 2.1.8 x86_64-darwin13.0)
```

**Comment by amaltson** [comment]
With enough quotes in places, I've been able get the syntax to be valid Ruby syntax, then only warnings appeared which are correctly disabled via `# rubocop:disable all`. The file now looks something like:

```ruby
#!/bin/sh

# rubocop:disable all
export PATH="/usr/local/bin:/usr/bin:$PATH"
exec ruby "-x" "$0" "$@"

#!ruby
# rubocop:enable all

puts ARGV[0]
# more regular Ruby code here
```

But the question above still stands I think. Thanks again!

**Comment by jonas054** [comment]
In order for RuboCop to reliably support shell script files that make us of Ruby's -x feature, it would have to parse the shell part of the file, which means adding support for parsing of one or more shell script languages. You might argue that supporting a few simple cases would be enough, but I don't think so. Request would come in for adding the ability to call `ruby -x` inside an `if`..`fi`, etc.

So I think your own solution, making the `sh` part syntactically valid Ruby code, is the way to go. Wanting RuboCop to parse other languages is asking too much.
